File: backend/app/schemas/template.py
"""Pydantic models for template management."""

from datetime import datetime
from typing import Optional

from pydantic import BaseModel, Field


class TemplateDetailOut(BaseModel):
    """Schema for template detail response."""

    template_name: str = Field(..., description="Template name")
    file_url: Optional[str] = Field(None, description="File URL")
    file_hvalue: Optional[str] = Field(None, description="File HValue")
    template_type: str = Field(..., description="Template type (text, media)")
    media_type: Optional[str] = Field(None, description="Media type (image, video)")
    uploaded_at: Optional[datetime] = Field(None, description="Upload timestamp")

    class Config:
        from_attributes = True


class TemplateCreateRequest(BaseModel):
    """Schema for creating a text template."""

    name: str = Field(..., description="Template name")
    language: str = Field(..., description="Template language code")
    category: str = Field(..., description="Template category")
    components: list = Field(..., description="Template components")


class TemplateSyncRequest(BaseModel):
    """Schema for syncing a template."""

    name: str = Field(..., description="Template name to sync")


class TemplateSendRequest(BaseModel):
    """Schema for sending a template."""

    template_name: str = Field(..., description="Template name")
    phone_numbers: Optional[str] = Field(None, description="Comma-separated phone numbers")
    basedon_value: Optional[str] = Field(None, description="Based on value (upload or Customer Base)")
    campaign_id: Optional[int] = Field(None, description="Campaign ID")

File: frontend/src/api/template.ts
import http from "./http";

export interface Template {
  id?: number;
  name: string;
  template_type?: string;
  templateType?: string;
  templateCreateStatus?: string;
  Status?: string;
  category?: string;
  media_type?: string;
}

export interface TemplateCreateRequest {
  name: string;
  language: string;
  category: string;
  components: Array<{
    type: string;
    format?: string;
    text?: string;
    example?: any;
  }>;
}

export interface TemplateSyncRequest {
  name: string;
}

export interface TemplateSendRequest {
  template_name: string;
  phone_numbers?: string;
  basedon_value?: string;
  campaign_id?: number;
}

export interface TemplateDetail {
  template_name: string;
  template_type: string;
  media_type?: string;
  file_url?: string;
  file_hvalue?: string;
  uploaded_at: string;
}

export interface TemplateListResponse {
  templates?: Template[];
  data?: Template[];
}

export async function getAllTemplates(): Promise<Template[]> {
  const response = await http.get<TemplateListResponse>("/campaign/templates/getAlltemplates");
  const data = response.data;
  console.log("API Response:", data);
  const templates = (data.templates || data.data || []) as Template[];
  console.log("Extracted templates:", templates);
  return templates;
}

export async function createTextTemplate(payload: TemplateCreateRequest): Promise<any> {
  return http.post("/campaign/templates/create-text-template", payload);
}

export async function createImageTemplate(formData: FormData): Promise<any> {
  return http.post("/campaign/templates/create-image-template", formData, {
    headers: {
      "Content-Type": "multipart/form-data",
    },
  });
}

export async function createVideoTemplate(formData: FormData): Promise<any> {
  return http.post("/campaign/templates/create-video-template", formData, {
    headers: {
      "Content-Type": "multipart/form-data",
    },
  });
}

export async function syncTemplate(payload: TemplateSyncRequest): Promise<any> {
  return http.post("/campaign/templates/sync-template", payload);
}

export async function getTemplateDetails(templateName: string): Promise<TemplateDetail> {
  return http.get(`/campaign/templates/${templateName}/details`);
}

export async function sendWhatsAppText(payload: TemplateSendRequest): Promise<any> {
  return http.post("/campaign/templates/sendWatsAppText", payload);
}

export async function sendWhatsAppImage(payload: TemplateSendRequest): Promise<any> {
  return http.post("/campaign/templates/sendWatsAppImage", payload);
}

export async function sendWhatsAppVideo(payload: TemplateSendRequest): Promise<any> {
  return http.post("/campaign/templates/sendWatsAppVideo", payload);
}

export async function downloadUploadTemplate(): Promise<Blob> {
  const response = await http.get<Blob>("/campaign/upload/template", {
    responseType: "blob",
  });
  return response.data;
}

